chandra_aca/dark_subtract.py

```diff
@@ -0,0 +1,289 @@
+        # Override the data_source to be explicit about maude source.
+        with fetch_sci.data_source("maude allow_subset=False"):
+            dat = fetch_sci.Msid("aacccdpt", start, stop)
+    else:
+        dat = fetch_sci.Msid("aacccdpt", start, stop)
+    return dat.vals, dat.times
+
+
+@retry.retry(exceptions=requests.exceptions.RequestException, delay=5, tries=3)
+def _get_dcsub_aca_images(
+    start: CxoTimeLike = None,
+    stop: CxoTimeLike = None,
+    aca_image_table=None,
+    t_ccd_vals=None,
+    t_ccd_times=None,
+    img_dark=None,
+    tccd_dark=None,
+    source="maude",
+):
+    """
+    Get dark current subtracted ACA images for a given time range.
+
+    This private method has expanded options for unit testing.
+    Users should use get_dcsub_aca_images instead.
+
+    Parameters
+    ----------
+    start : CxoTimeLike, optional
+        Start time of the time range. Default is None.
+    stop : CxoTimeLike, optional
+        Stop time of the time range. Default is None.
+    aca_image_table : astropy table of ACA images, optional
+
+    t_ccd_vals : array-like, optional
+        Array of CCD temperatures. Default is None.
+    t_ccd_times : array-like, optional
+        Array of CCD temperature times. Default is None.
+    img_dark : array-like, optional
+        Dark current image. Default is None.
+    tccd_dark : float, optional
+        Dark current CCD temperature. Default is None.
+    source : str, optional
+        Source of the ACA images. Can be 'maude' or 'mica'. Default is 'maude'.
+
+    Returns
+    -------
+    astropy.table.Table
+        Table of aca image data including raw and dark current subtracted images.
+        Raw images are in column 'IMG', dark current subtracted images are in column 'IMGBGSUB',
+        dark current cut-out images are in column 'IMGDARK', and interpolated
+        CCD temperature values are in column 'AACCCDPT'.
+
+    """
+
+    if start is None and aca_image_table is None:
+        raise ValueError("start must be defined if aca_image_table is not defined")
+    if stop is None and aca_image_table is None:
+        raise ValueError("stop must be defined if aca_image_table is not defined")
+
+    if aca_image_table is not None:
+        # If supplied an image table, use that, but confirm it has the right pieces.
+        assert type(aca_image_table) is Table
+
+        # Require aca_image_table have columns IMG, IMGROW0_8X8, IMGCOL0_8X8, TIME.
+        # If this is a table of mica data that doesn't have those columns, see
+        # mica.archive.aca_l0.get_aca_images for how to get the data or use that method.
+        assert "IMG" in aca_image_table.colnames
+        assert "IMGROW0_8X8" in aca_image_table.colnames
+        assert "IMGCOL0_8X8" in aca_image_table.colnames
+        assert "TIME" in aca_image_table.colnames
+        # require that IMGROW0_8X8 and IMGCOL0_8X8 are masked columns
+        assert hasattr(aca_image_table["IMGROW0_8X8"], "mask")
```

Comment:
I'm curious what breaks if the images are not masked? Personally I really do not like masked arrays if the data do not require masks, because the `repr` is so ugly.

Reply:
I was just following your guidance to be explicit and define what we want for inputs instead of handling every case.  In this case, because I've assumed the masked array in the input (which was what I was getting from maude_decom.get_aca_images) this code really just needs to get the unmasked elements or fill the masked ones (I chose to fill).

Reply:
Unless I am missing something, the `get_aca_images()` output from maude and L0 (mica) cannot ever have masked elements for IMGROW/COL0_8X8 or TIME. This is because both of those only return complete images and by definition a complete image has a row, column, time, and image data. The image can be masked of course, though only with fixed 4x4 or 6x6 mask patterns.

So I do not see any reason to complicate this code with ad hoc treatment of masked values. You can hit the input requirements like this:
```
for name in ["IMG", "IMGROW0_8X8", "IMGCOL0_8X8", "TIME"]:
    col = aca_image_table[name]  # raises KeyError if not there
    if name != "IMG" and hasattr(col, "mask") and np.any(col.mask):
        raise ValueError(f"masked values in column {name} are not allowed")
```
There is nothing in the downstream code that strictly depends on this being an astropy Table. It could be a numpy structured array, or even a dict.

Reply:
I had ad-hoc treatment of masked values because I think staggered_aca_slice will not work on a masked array even if there are no masked values.  If we decide to unmask with something like https://github.com/sot/chandra_aca/pull/182 that's fine.  I just need a plan.